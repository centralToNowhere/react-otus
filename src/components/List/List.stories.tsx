import React from "react";
import { Story, Meta } from "@storybook/react";
import { IListProps, List } from "@/components/List/List";
import { l10n } from "@/l10n/ru";

export default {
  title: "Components/List",
  component: List,
  args: {
    items: [],
  },
} as Meta;

const Template: Story<IListProps> = (args) => <List {...args} />;

export const ListDefault = Template.bind({});

ListDefault.args = {
  items: [
    {
      content: l10n.rule3,
      items: [
        {
          content: l10n.rule3_1,
        },
        {
          content: l10n.rule3_2,
        },
      ],
    },
  ],
};
