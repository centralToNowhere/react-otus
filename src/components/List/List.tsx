import React from "react";
import styled from "@emotion/styled";

interface IListItem {
  content: string;
  items?: IListItem[] | null;
}

export interface IListProps {
  items: IListItem[];
}

export class List extends React.Component<IListProps> {
  constructor(props: IListProps) {
    super(props);
  }

  getKey(string: string) {
    // assumes that all strings in list are unique
    const res = [...string].reduce((acc, curr) => {
      acc = (acc * curr.charCodeAt(0)) / 1000;
      return acc;
    }, 1);

    return res.toString();
  }

  getContent = (item: IListItem) => {
    return (
      <React.Fragment key={this.getKey(item.content)}>
        <li>{item.content}</li>
        {item.items ? (
          <ul>
            {item.items.map((item) => {
              return this.getContent(item);
            })}
          </ul>
        ) : null}
      </React.Fragment>
    );
  };

  render() {
    return (
      <StyledList>
        <ul>{this.props.items.map(this.getContent)}</ul>
      </StyledList>
    );
  }
}

const StyledList = styled.div`
  padding: 2em;

  li {
    font-size: 1.5rem;
    margin-bottom: 1em;
    list-style: disc;
  }
`;
