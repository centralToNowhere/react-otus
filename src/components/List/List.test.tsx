import React from "react";
import { render, screen } from "@testing-library/react";
import { List } from "@/components/List/List";

describe("List tests", () => {
  it("should render list", () => {
    const test = [
      {
        content: "test",
      },
      {
        content: "test2",
        items: [
          {
            content: "test3",
            items: [
              {
                content: "test4",
              },
            ],
          },
        ],
      },
    ];

    const { asFragment } = render(<List items={test} />);

    const test1 = screen.getByText("test");
    const test2 = screen.getByText("test2");
    const test3 = screen.getByText("test3");
    const test4 = screen.getByText("test4");

    expect(test1).toBeInTheDocument();
    expect(test1.nextElementSibling).toContainElement(test2);

    expect(test2).toBeInTheDocument();
    expect(test2.nextElementSibling).toContainElement(test3);

    expect(test3).toBeInTheDocument();
    expect(test3.nextElementSibling).toContainElement(test4);

    expect(test4).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });
});
