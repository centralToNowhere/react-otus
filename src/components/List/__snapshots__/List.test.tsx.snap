// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List tests should render list 1`] = `
<DocumentFragment>
  .emotion-0 li {
  margin-bottom: 0.5em;
  list-style: disc;
}

<div
    class="list emotion-0 emotion-1"
  >
    <ul>
      <li>
        test
      </li>
      <li>
        test2
      </li>
      <ul>
        <li>
          test3
        </li>
        <ul>
          <li>
            test4
          </li>
        </ul>
      </ul>
    </ul>
  </div>
</DocumentFragment>
`;
